agent/ecr/factory.go

```diff
@@ -15,74 +15,59 @@
 package ecr
 
 import (
+	"fmt"
 	"net/http"
-	"sync"
 	"time"
 
-	"github.com/aws/amazon-ecs-agent/agent/async"
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/credentials"
 	ecrapi "github.com/aws/amazon-ecs-agent/agent/ecr/model/ecr"
 	"github.com/aws/amazon-ecs-agent/agent/httpclient"
 	"github.com/aws/aws-sdk-go/aws"
+	awscreds "github.com/aws/aws-sdk-go/aws/credentials"
 	"github.com/aws/aws-sdk-go/aws/session"
 )
 
 type ECRFactory interface {
-	GetClient(region, endpointOverride string) ECRClient
+	GetClient(*api.ECRAuthData) (ECRClient, error)
 }
 
 type ecrFactory struct {
 	httpClient *http.Client
-
-	clientsLock sync.Mutex
-	clients     map[cacheKey]ECRClient
-}
-
-type cacheKey struct {
-	region           string
-	endpointOverride string
 }
 
 const (
 	roundtripTimeout = 5 * time.Second
-	tokenCacheSize   = 100
-	tokenCacheTTL    = 12 * time.Hour
 )
 
 // NewECRFactory returns an ECRFactory capable of producing ECRSDK clients
 func NewECRFactory(acceptInsecureCert bool) ECRFactory {
 	return &ecrFactory{
 		httpClient: httpclient.New(roundtripTimeout, acceptInsecureCert),
-		clients:    make(map[cacheKey]ECRClient),
 	}
 }
 
-// GetClient returns the correct region- and endpoint-aware client
-func (factory *ecrFactory) GetClient(region, endpointOverride string) ECRClient {
-	key := cacheKey{region: region, endpointOverride: endpointOverride}
-	client, ok := factory.clients[key]
-	if ok {
-		return client
+// GetClient returns the correct region and endpoint aware client
+func (factory *ecrFactory) GetClient(authData *api.ECRAuthData) (ECRClient, error) {
+	cfg := aws.NewConfig().WithRegion(authData.Region).WithHTTPClient(factory.httpClient)
+	if authData.EndpointOverride != "" {
+		cfg.Region = aws.String(authData.EndpointOverride)
 	}
 
-	factory.clientsLock.Lock()
-	defer factory.clientsLock.Unlock()
-	client, ok = factory.clients[key]
-	if ok {
-		return client
+	if authData.UseExecutionRole {
+		if authData.GetPullCredentials() == (credentials.IAMRoleCredentials{}) {
```

Comment:
why wrap the object with `()` ?

Reply:
without this, gofmt will return an error and won't compile. I will get rid of this with `utils.zeroornil`.